### ISSUE TITLE
Variable description in list_census_vectors

### ISSUE THREAD

**Issue by dshkol** [mountainMath/cancensus#97]
This is a more general look at the issue identified in #87 

The process for creating the variable descriptions should work roughly something like this:

1. Take list of vectors and identify the appropriate paren vector as specified in the `parent_vector` field
2. Look up that parent_vector's label and prepend it to the child vector's label
3. Do this recursively to pull in the grand-parent vector, the grand-grand-parent vector, and so on, as necessary. The 2016 census has up to 6 levels of depth, but I don't want to hard code that. 

The current process to traverse the label hierarchy to create concatenated variable descriptions works like this, where `result` is a data frame of that census vectors and labels for that census dataset:

```
# traverse hierarchy to add description field to variables
    result$description <- result$label
    list <- result
    browser()
    while (any(!is.na(list$parent_vector))) {
      parent_list=result %>% dplyr::filter(vector %in% list$parent_vector)
      result$description[!is.na(list$parent_vector)] <-
        paste(parent_list$label,result$description[!is.na(list$parent_vector)],sep=", ")
      list=parent_list
    }
```

This doesn't actually work correctly, as it doesn't index the positions of the parent vectors correctly, and results in misaligned concatenation. 

@mountainMath and myself have tried a few different approaches over the last while to fix this, but have been stumped. I've had better luck by focusing on recursive join functions but can't quite get it work.

As an alternative, I put together an alternative approach that uses a function that traverses the label hierarchy for any given vector, and then splitting the data frame and vectorizing over that list. It works like this:

```
# traversal function
parent_labels <- function(vector_list) {
  base <- vector_list
  n <- 0
  vector_list <- result[result$vector == base$parent_vector,] %>% distinct(vector, .keep_all = TRUE)
  # Recursively assemble all parents of any vector
  while(n != nrow(vector_list)) {
    n = nrow(vector_list)
    new_list <- result %>% filter(vector %in% vector_list$parent_vector)
    vector_list <- vector_list %>% rbind(new_list) %>% distinct(vector, .keep_all = TRUE)
  }
  # Reverse order and collapse the parent labels into a single string
  labels <- vector_list[order(desc(row.names(vector_list))),]$label
  labels <- paste(labels, collapse = ": ")
  return(labels)
}

result_split <- split(result,seq_len(nrow(result)))
full_labels <- lapply(result_split, parent_labels)

```

Despite using a vectorized approach (*apply family is generally the fastest way to do anything in R), this process is unacceptably slow when run on the entire dataset. I've benchmarked each function call at around 0.015 seconds, but if you have 5000 vectors, then that's too slow to be used reliably. 

IMO, this is the biggest outstanding issue left before cleaning up for CRAN. The vectors are hard to find and hard to parse without reliable descriptions, and this affects `search_census_vectors` as well. Does anyone have a better solution for addressing this?

**Comment by atheriel** [comment]
To be honest I don't think this should be a hard requirement before submission to CRAN. While I do think that discoverability is important, I also think that "works ok" is alright for a first release, especially on this topic. Hierarchies are hard, and I think people get that to a degree.

I will take a crack at improving this function at some point soon, though!

**Comment by dshkol** [comment]
While I agree that maybe we don't need to have this fixed before pushing to CRAN, I disagree that it works ok - because the descriptions are in many cases completely wrong. We could scrap the description column in the meanwhile altogether, but that would make it harder to search for vectors at times as the labels alone arent always descriptive enough to search on, and in some cases are not unique.

**Comment by atheriel** [comment]
That's fair.

**Comment by mountainMath** [comment]
I have a server-side fix for this, that also has the advantage of adding a couple of extra markers that help with structuring the data. That is placeholder names that don't correspond to actual census variables but help structure the hierarchy as in the CensusMapper variable selection tree. Once the current set of changes are merged I will push this live on CensusMapper and make the appropriate change in `cancensus` function.

**Comment by dshkol** [comment]
I think the only thing that needs to be merged in is #103 before making this change. Documentation and pkgdown content will need to be regenerated after this change as there will be some small tweaks the documentation on `list_census_vectors()` and `search_census_vectors()`.

**Comment by dshkol** [comment]
This has been fixed on the server side and implemented in retrieval in #104
